Adding 3D mortar capability

jbadger95 opened the pull request:
Addresses #13080 "Extend to 3D" task.

A segment-based mortar implementation following the [thesis](https://mediatum.ub.tum.de/1109994) of Alexander Popp. The implementation is largely compatible with the 2D implementation, differing mostly in creation of the mortar segment mesh.

@lindsayad

Commit:
Initial commit of class for computing mortar segment mesh elements in 3D.

Currently the class computes and outputs the polygon clipping and triangulation of secondary elements against each primary element individually. However, this does not allow for sections of secondary elements without a corresponding primary element to be triangulated. This class is therefore currently restricted to problems where the primary interface face projection covers the secondary interface.

Commit:
Added 3D mortar generation routine, temporarily removed various restrictive 1D mortar tests

Commit:
Added routine for computing 3D mortar segment meshes

Commit:
Initial modifications for 3D mortar constraints.

This enables initial 3D mortar tests that produce reasonable results. This is still unrefined and will require modifications but is being commited to archive a first minimally functional 3D mortar implementation.

Commit:
Removing old cant-do-3d test and replacing it with a periodic mortar test on same mesh.

Commit:
Added problem input file for enforcing continuity across two cube domains with non-conforming discretizatiosn

Commit:
Cleaned up mortar functor and added documentation comments to MortarSegmentHelper

Commit:
Added support for mortar segment elements without corresponding primary element and introduced fudge-factor for msm elements with partial overlap.

Also cleaned up MortarSegmentHelper including:
- Made tolerances dimensionally consistent
- Added comments about tolerancing intersections

refs #13080

Commit:
Remove banned words

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/768394/) on ad9acc3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 71ac6fd78e759ddcb10e047c2b8aefdb8b7b3cd5`

dschwen commented on framework/src/loops/ComputeMortarFunctor.C:
```diff
@@ -123,17 +130,38 @@ ComputeMortarFunctor::operator()()
     // Compute a JxW for the actual mortar segment element (not the lower dimensional element on
     // the secondary face!)
-    _subproblem.reinitMortarElem(msm_elem);
+    // fudge_factor is a convenient way of handling seconary elements with only partial overlapping
```
```suggestion
    // fudge_factor is a convenient way of handling secondary elements with only partial overlap
```

jbadger95 replied:
Thanks, I'm curious to know if you have any thoughts about this work around? 

I'm trying to think of some tests to make sure we still get the expected rate of convergence but its a bit tricky. For example, let's say we have 2 cubes next to each other with an equal value constraint on the 2 closest sides and the 2 cubes are slightly offset so that some of the elements on the secondary side aren't fully covered by elements on the primary side. If we use MMS, all the boundaries where the constraint isn't will essentially have fluxes fixed, but on those elements on the secondary side without  corresponding primary elements we'll be enforcing a 0 flux when they shouldn't be.

I can work on implementing this case more carefully (by actually forming the mortar segment mesh on these elements) but it'll complicate my implementation more so I'm hoping this is good enough.

jbadger95 replied:
Just to close this discussion, the correct way to deal with mortar segments without a primary element across is to simply not assemble them. All has_primary logic has been removed and mortar segments without a primary element across are not included in the mortar segment mesh. This does create a conditioning issue when some elements only have very small contributions but will be addressed in the future with an update to dual shape functions calculations.

Commit:
Clang formatting

Commit:
Fixed build errors (oops)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/768744/) on 659c7e3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b9fe20a441fafce9b34bdb74f2c8095068cbb575`

Commit:
clang formatting

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/768788/) on 5652944 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18203/site)

This comment will be updated on new commits.

dschwen commented:
@jbadger95 you're probably already aware of this thesis: https://mediatum.ub.tum.de/doc/1363363/1363363.pdf 
Am I right that the current implementation is surface/surface contact? How difficult would the extension to edge/surface, edge/edge,  corner/surface be?
@recuero can comment on the partial cover issue when he's back. I remember him explaining a potential approach (which I think amounted to hybrid mortar / node-on-face... but don't nail me down on that).

jbadger95 commented:
@dschwen yes this implementation mostly follows Popp's thesis, it was one of the resources @recuero recommended but I haven't been able to find a reference on this case of when there is no primary element across in a tying problem. @lindsayad and I have observed some interesting behavior in this case (it gives rise to a discontinuous LM and the dual basis functions often used for stability show some interesting behavior in trying to reproduce the discontinuity) so I'd be interested to know if anyone knows of a reference. 

This implementation is for surface/surface contact. To be honest I'm not sure how difficult those other cases would be, I just started learning about mortar methods a few weeks ago. I'm curious though, are there any specific use cases you had in mind for edge/surface and corner/surface?

lindsayad commented:
> it gives rise to a discontinuous LM and the dual basis functions often used for stability show some interesting behavior in trying to reproduce the discontinuity

interesting behavior = oscillatory behavior = bad behavior haha. But my hypothesis is that we could fix this by forcing the LM to 0 on a per-dof basis like we do for dofs in mechanical contact when the secondary doesn't project onto the primary.

Commit:
Added convergence study, fixed mortar segment info dimension check that was previously disabling 3D mortar

recuero commented:
@dschwen: @jbadger95 is going for the most accurate approach from the start, no need for approximations based on projections that neglect forming the segments appropriately. One advantage of the less accurate approach is compute time according to some authors, but I think MOOSE does a pretty efficient job building the segments, so that "hybrid" approach not needed at this point IMO.

Commit:
Initial commit of second order 3D mortar.

For second order mortar methods the mortar segment mesh is linearized and still composed of 1st order segments.

In this implementation, second order elements are simply sub-divided into four sub-elements which are linearized and the first-order algorithm is applied.

Implmentation is functional and realizes expected 3rd order primal and 2nd order convergence rates for enforcing continuity on non-matching meshes but is not yet complete.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/772633/) on e595f23 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f59e5424802e6b294e602ea4597ce6e78bfa08ed`

Commit:
Clang formatting

Commit:
Added documentation for mortar segment mesh generaton

Commit:
Added citation and minor edits to automatic mortar generation

Commit:
Adding mortar problem on curved (sphere) mesh.

Note second order mortar on curved meshes doesn't work yet, need to project to intermediate first order mesh first, then transfer to second order mesh

Commit:
Added second order convergence study

Commit:
Added second order mortar capability on curved meshes

Quadrature points now projected back to first order elements then transferred to second order elements. Current implementation is a bit clumsy but demonstrates expected convergence rates on curved meshes

Commit:
Added more extensive comments to mortar segment generation routines, cleaned up various aspects

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/778061/) on e8824c8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

Commit:
Clang formatting

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 577b8fbe2d31c98e8220d79569694c5edb93b8b4
+Subproject commit c02af017673655633edfc4ebf8454f46bd94f059
```
**Caution!** This contains a submodule update

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ffab5c914090ad625645faa1bb4251d6b46f1c26
+Subproject commit c02af017673655633edfc4ebf8454f46bd94f059
```
**Caution!** This contains a submodule update

jbadger95 committed:
Added numerous tests for 3D mortar continuity constraints

jbadger95 committed:
Added option to change quadrature rule on mortar segment mesh

jbadger95 committed:
Cleaned up mortar segment mesh elements, added msm statistics and mesh printing, and non-dimensionalized tolerance in ComputeMortarFunctor that was dropping small elements

jbadger95 committed:
Updated documentation to include 3D mortar and expanded documentation on mortar segment mesh generation

jbadger95 committed:
fixed compute mortar functor errors for 2d mortar tests

jbadger95 committed:
Rebase on next

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/793179/) on 2351f52 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9b4989eac6fdf5845ea5b32fbc9a18386fddcb42`

jbadger95 committed:
Clang formatting

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/793466/) on aeb9346 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9b4989eac6fdf5845ea5b32fbc9a18386fddcb42`

jbadger95 committed:
Added logic for finding inactive LM DoFs. These DoFs will simply be set to zero in mortar constraint objects

jbadger95 committed:
Removed has_primary logic from base constraints and added Zeroing Inactive DoFs

jbadger95 committed:
Modifyied contact modules to use weighted traction (LM) on non-dual basis. Removed has_primary logic.

jbadger95 committed:
Removed has primary logic from remaining contact objects

jbadger95 committed:
Removed has_primary logic from gap conductance constraints

jbadger95 committed:
Added tolerancing for dropping secondary elements with only small primary contributions

jbadger95 committed:
Removed logic for dropping elements with only small primary contribution. This leads to oscillations since dropped dofs are strongly set to zero but are still coupled to active DoFs. Better solution is to use dual shape functions for edge dropping case, but work still needed to correctly handle this case.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/812133/) on dac7578 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 355da44a1976c3b9d9326737a3ca392cc5fd9c1f`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/812164/) on a9c85c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 355da44a1976c3b9d9326737a3ca392cc5fd9c1f`

jbadger95 committed:
Clang Formatting

jbadger95 committed:
Fixing compiler warnings for possible uninitialized variables

jbadger95 committed:
Fixed error with multiple primary nodes try to project onto same point on secondary surface

jbadger95 committed:
Updated weighted friction computation to align with Gitterle paper, updated some contact weighting constants to reflect new scaling weighted instead of nodal LM values

jbadger95 committed:
reverted to nodal contact constraints, added option for normalizing c

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/814810/) on e74c01e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e3c48025ba7d3169cd5de39de0fcfd25ca443aca`

jbadger95 committed:
Updated tests. Gap conductance regolded due to new handling of edge dropping.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/815572/) on 62c174c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9326e72914620ea18afaf58a3bf3d28451be9a40`

jbadger95 committed:
Updating cylinder friction contact test for new geometry (secondary interface limited to avoid edge dropping cases)

jbadger95 committed:
Re-golded normal nodal test, fixed periodic test naming.

jbadger95 committed:
Updated documentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/817185/) on 95500a6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format abd1920d512a62749ba656f7a0278c6ce2d977ca`

jbadger95 committed:
Added mesh files for 3d mortar convergence tests

jbadger95 committed:
Added give_me_wrong_results parameter to enable old handling of edge dropping, reverted regolded tests accordingly

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/822150/) on fa9f0e8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b42ecd3a731e49a38c8cb5f154d628e771ac5e4`

jbadger95 committed:
Added give_me_wrong_results = false option to normal-nodal-lm-tan-tolerance test, this is necessary any time mortar and nodal constraints are mixed. This is a bit confusing but the hope is that the old edge dropping behavior will be depricated in the near future anyway.

jbadger95 committed:
Re-golded normal-nodal-lm-tan test for edge dropping. When nodal and LM contact constraints combined give_me_wrong_results needs to be set to false

jbadger95 committed:
Fixed give_me_wrong_results behavior (detection of inactive nodes was wrong and in 3D the computation to detect dropping was wrong). Also updated contact objects so that strongly zeroed DoFs will not be coupled to the system.

recuero commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -146,25 +166,47 @@ ComputeWeightedGapLMMechanicalContact::computeJacobian(const Moose::MortarType m
 
 void
-ComputeWeightedGapLMMechanicalContact::enforceConstraintOnNode(const Node * const node)
+ComputeWeightedGapLMMechanicalContact::wrongPost(const std::unordered_set<const Node *> & inactive_lm_nodes)
```
`wrongPost`, could be more descriptive.

jbadger95 replied:
It's a copy of post that skips assembling any inactive nodes when the old edge dropping behavior is enabled, I was just trying to keep with naming these things like their counterpart routines but with `wrong` to point out they correspond to the `give_me_wrong_results` option. 

I can try to think of something more descriptive but they're supposed to be deprecated soon anyway

recuero commented on modules/contact/test/tests/pdass_problems/hertz_cyl_half_1deg.jou:
```diff
@@ -88,7 +88,7 @@ mesh surf all
 
```
Does this journal file correspond to the mesh used in the test?

jbadger95 replied:
You're right I reverted the mesh but not the journal file, good catch!

recuero commented on test/tests/mortar/3d-periodic/periodic.i:
```diff
@@ -59,7 +60,7 @@
     primary_subdomain = 12
     secondary_subdomain = 11
-
+    delta = 0.1
```
IIRC This is a stabilization parameter: can the name be more descriptive?

jbadger95 replied:
This is the stabilization parameter but that's not something I implemented. I could change it but that might mess up people's tests and applications since it's been around for a while

recuero replied:
Ok then, never mind

recuero commented on test/tests/mortar/3d-periodic/tests:
```diff
@@ -0,0 +1,12 @@
+    input = 'periodic.i'
+    exodiff = 'periodic_out.e'
+    requirement = 'We shall be able to enforce periodic boundary conditions on 3d domains using mortar constraints.'
+    map = False
+    partial = True
```
Why these additional settings?

recuero commented on modules/contact/include/constraints/ComputeWeightedGapLMMechanicalContact.h:
```diff
@@ -72,10 +80,16 @@ class ComputeWeightedGapLMMechanicalContact : public ADMortarConstraint
   /// The value of the gap at the current quadrature point
   ADReal _qp_gap;
+  /// The value of the LM at the current quadrature point
+  Real _qp_factor;
```
Are you adding mechanical contact tests for 3D?

jbadger95 replied:
I haven't added any, I can look at adding a couple simple ones but that's definitely something I could have done better.

recuero reviewed (commented):
This looks great to me. A lot of progress and comments explaining every step!

jbadger95 commented on framework/include/base/Assembly.h:
```diff
@@ -598,6 +607,13 @@ class Assembly
+  /**
+   * reinitialize a mortar segment mesh element in order to get a proper JxW
+   */
+  void reinitMortarElem(const Elem * elem,
```
I was used for my old treatment of edge dropping for 3d mortar (I was modifying the weights), it's not used now though so I'll remove it.

jbadger95 commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -67,7 +67,8 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
                             bool on_displaced,
-                            bool periodic);
+                            bool periodic,
+                            bool give_me_wrong_results);
```
Yeah, I agree it could be more descriptive, ideally this whole option will be removed soon though. The remaining changes to make correct treatment edge dropping robust really aren't that huge, I just ran out of time. I'll change it.

jbadger95 replied:
The naming scheme has been changed!

jbadger95 commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -179,6 +198,19 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+   * Used to enforce zero values on inactive DoFs of nodal variables.
+   */
+  void computeInactiveLMNodes();
+  void computeWrongInactiveLMNodes();
```
All these routines with wrong in the name are to enable the old edge dropping behavior (originally they weren't so extensive but the current treatment is exactly the same for contact as what was implemented before). I'll add some comments but the hope is for these to be removed eventually so I wasn't as careful

jbadger95 commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -209,6 +241,26 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+   */
+  const std::unordered_set<const Node *> & getInactiveLMNodes() const
+  {
+    return inactive_local_lm_nodes;
```
In this case we're getting the actual geometric nodes that should be inactive, later on in MortarConstraintBase we use the system and variable information to get the actual degrees of freedom. For example in frictional contact each node will have a normal and frictional DoF associated with it. The other routine getInactiveLMElems does a similar thing but for elemental variables, we're just computing the geometric entities so that later we can find all the DoFs that we need to set to zero.

jbadger95 commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -290,6 +342,12 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
   // Container for storing the nodal normal vector associated with each secondary node.
   std::unordered_map<const Node *, Point> secondary_node_to_nodal_normal;
 
+  // List of inactive lagrange multiplier nodes (for nodal variables)
+  std::unordered_set<const Node *> inactive_local_lm_nodes;
+
+  // List of inactive lagrange multiplier nodes (for elemental variables)
+  std::unordered_set<const Elem *> inactive_local_lm_elems;
+
```
Yes, the LM lives on the secondary surface so these are all nodes and elems corresponding to the secondary surface

jbadger95 commented on framework/include/constraints/MortarSegmentHelper.h:
```diff
@@ -0,0 +1,154 @@
+      const Point & p1, const Point & p2, const Point & q1, const Point & q2, Real & s) const;
+
+  /**
+   * Check that a point is inside the secondary polygon (for verification only)
```
I tried to use elem to talk about the faces when they were in 3d, and polygon to talk about them when they had been projected onto a 2D plane. So this is just a secondary face element that has been projected onto a 2D plane.

jbadger95 commented on framework/include/constraints/MortarSegmentHelper.h:
```diff
@@ -0,0 +1,154 @@
+   */
+  std::vector<Point> clipPoly(const std::vector<Point> & primary_nodes) const;
+
+  /**
+   * Prints python plotting commands (to terminal) for visualizing polygons (for debugging only)
+   */
+  void plotPoly(const std::vector<Point> & poly) const;
+
+  /**
+   * Prints python plotting commands (to terminal) for visualizing triangulation (for debugging
+   * only)
+   */
+  void plotTriangulation(const std::vector<Point> & nodes,
+                         const std::vector<std::vector<unsigned int>> & elem_to_nodes) const;
+
```
They aren't as necessary anymore, if you set _debug=true in AutomaticMortarGeneration it will output the mortar segment mesh as an exodus file. This can be useful for debugging the polygon clipping and stuff but I hope I've managed to make this pretty robust (it took a few iterations)

recuero replied:
Right. Only need to switch that flag in the constructor IIRC

jbadger95 commented on framework/include/constraints/MortarSegmentHelper.h:
```diff
@@ -0,0 +1,154 @@
+                       std::vector<std::vector<unsigned int>> & tri_map) const;
+
+  /**
+   * Get mortar segments generated by a secondary and primary node pair
```
My mistake I meant elem pair, I'll modify that

jbadger95 commented on framework/include/constraints/MortarSegmentHelper.h:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * Tolerance for intersection and clipping
+   */
+  Real _tolerance = 1e-8;
```
I don't think so, it really shouldn't be changed (it's not as delicate now that I've modified how the intersection is computed). It's used in a few places like to help with the conditioning of the getIntersection routine (which is horribly conditioned when liines are parallel like when dealing with conforming meshes). Before I put this `s` business into the calculation, when the problem was poorly conditioned it would spit out random points that weren't even on my surfaces. The tolerancing is less important for the intersection now that it's impossible for it to return points not on the polygon but it also plays a role in limiting the proliferation of nodes (before it would produce about 4x as many nodes as needed for conforming meshes which in turn created a bunch of tiny mortar segments). 

It works well now, I haven't any problems and have tested it thoroughly but this clipping stuff is one part of the mortar code I would avoid modifying, it looks benign but it took multiple iterations to get working robustly.

jbadger95 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1352,7 +1352,8 @@ class FEProblemBase : public SubProblem, public Restartable
       bool on_displaced,
-      bool periodic);
+      bool periodic,
+      bool give_me_wrong_results);
```
Yeah it's odd, I needed the parameter during the construction of the mortar segment mesh though and this object calls the routines that construct it. Maybe there's a way to put it in MortarConstraintBase but with the inheritance structure of these objects I didn't see a straightforward way to define it in MortarConstraintBase and access it in AutomaticMortarGeneration. I'm not sure I see a better way to do it, I'm still fairly new to C++ though so maybe I'm missing something?

jbadger95 commented on framework/src/base/Assembly.C:
```diff
@@ -697,6 +701,31 @@ Assembly::setNeighborQRule(QBase * qrule, unsigned int dim)
+  if (order != _qrule_msm->get_order())
+  {
+    // If custom mortar qrule has not yet been specified
+    if (!_custom_mortar_qrule)
+    {
+      _custom_mortar_qrule = true;
+      const unsigned int dim = _qrule_msm->get_dim();
+      const QuadratureType type = _qrule_msm->type();
+      delete _qrule_msm;
+
+      _const_qrule_msm = _qrule_msm = QBase::build(type, dim, order).release();
+      _fe_msm->attach_quadrature_rule(_qrule_msm);
```
The mortar quadrature rule is defined when Assembly is initialized and before the MortarConstraint objects exist, so if a different one is specified in the MortarConstraint this will just swap a new rule in for the old one.

recuero commented on framework/src/base/Assembly.C:
```diff
@@ -697,6 +701,31 @@ Assembly::setNeighborQRule(QBase * qrule, unsigned int dim)
+    if (!_custom_mortar_qrule)
+    {
+      _custom_mortar_qrule = true;
+      const unsigned int dim = _qrule_msm->get_dim();
```
> For 2D I can't think of a reason why we would want to use anything other than default but I guess I don't see any harm in allowing users to do it if they want to?
👍

jbadger95 committed:
Address Antonio review comments part 1

lindsayad reviewed (changes requested):
Really impressive PR. I love your code style

jbadger95 commented on framework/src/base/Assembly.C:
```diff
@@ -162,7 +163,9 @@ Assembly::Assembly(SystemBase & sys, THREAD_ID tid)
-  _fe_msm = FEGenericBase<Real>::build(_mesh_dimension - 1, FEType(helper_order, LAGRANGE));
+  _fe_msm = (_mesh_dimension == 2)
+                ? FEGenericBase<Real>::build(_mesh_dimension - 1, FEType(helper_order, LAGRANGE))
+                : FEGenericBase<Real>::build(_mesh_dimension - 1, FEType(FIRST, LAGRANGE));
```
The mortar segments are always TRI3, I think this is just used to define the quadrature points IIRC.

lindsayad replied:
Ah that makes sense. Would you mind adding a comment stating add that?

jbadger95 replied:
Done!

jbadger95 commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -331,6 +343,11 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
     if (current_mortar_segment == nullptr)
       mooseError("Unable to find appropriate mortar segment during linear search!");
 
+    // // If node lands on endpoint of segment, don't split
+    // info = &msm_elem_to_info.at(current_mortar_segment);
+    // if (info->xi1_a == xi1 || xi1 == info->xi1_b)
+    //   continue;
+
```
This was part of the fix to that one guy's 2D mortar problem with the funky geometry. For that fix we also change the signs a couple lines above in `(info->xi1_a < xi1 && xi1 < info->xi1_b)` to `<=`. This got taken out while I was trying to diagnose those ironing issues in 2D (where I was using the invalid quadratures), it could be put back in if you with but it was a very funky case that probably doesn't need to be handled.

lindsayad replied:
Hmm in that case, let's either remove the lines and explain what they were for in the commit message, or add another comment about what this might be used for (e.g. a weird geometry with duplicate notes ... that's what that person had right?)

jbadger95 replied:
I put it back in, is the original comment not descriptive enough (that the element won't be split if the point lands on an endpoint)? Yeah it was a problem of two nodes projecting onto the exact same point.

jbadger95 commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -568,11 +585,43 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
     mortar_segment_set.insert(new_elem_right);
   }
 
-  // Set up the the mortar segment neighbor information.
-  mortar_segment_mesh->allow_renumbering(true);
-  mortar_segment_mesh->skip_partitioning(true);
-  mortar_segment_mesh->allow_find_neighbors(true);
-  mortar_segment_mesh->prepare_for_use();
```
Sorry @recuero, didn't see this till just now (didn't expand this file). Yeah essentially I removed them because I thought they were unnecessary since we're using this mesh as more of a holder for mortar segments but it probably doesn't save much. I'll put them back in.

jbadger95 replied:
Added back in to both the 2D and 3D mortar segment mesh generation

jbadger95 commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -568,11 +585,43 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
+    MortarSegmentInfo & msinfo = msm_elem_to_info.at(msm_elem);
+    Elem * primary_elem = const_cast<Elem *>(msinfo.primary_elem);
+    if (primary_elem == nullptr || std::abs(msinfo.xi2_a) > 1.0 + TOLERANCE ||
+        std::abs(msinfo.xi2_b) > 1.0 + TOLERANCE)
```
It's not, this tolerance is just the default tolerance, it's maybe not necessary, it's really just checking that these values are valid (the ironing problem I was -99, the invalid value so this is just to make sure the segment is a valid segment)

jbadger95 commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -608,17 +657,736 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
+AutomaticMortarGeneration::buildMortarSegmentMesh3d()
+{
+  // Add an integer flag to mortar segment mesh to keep track of which subelem
+  // of second order primal elements mortar segments correspond to
```
second order, this has do to with the subdivision of second order elements into first order subelements

jbadger95 committed:
Addressing Alex comments part 1

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented on test/tests/mortar/convergence-studies/continuity-3d/tests:
```diff
@@ -0,0 +1,64 @@
+    method = '!dbg'
+    required_python_packages = 'sympy pandas matplotlib'
+    mumps = 'TRUE'
+    heavy = True
```
Ok yea, then the heavy designation definitely makes good sense

jbadger95 committed:
Addressing Alex and Antonio comments part 2

jbadger95 commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -608,17 +657,736 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
+        normal.unit();
+
+        // Build and store linearized sub-elements for later use
+        mortar_segment_helper[sel] = new MortarSegmentHelper(nodes, center, normal);
```
Done!

jbadger95 commented on framework/src/constraints/MortarSegmentHelper.C:
```diff
@@ -0,0 +1,378 @@
+  //     tri_map.push_back({1 + offset, 2 + offset, 3 + offset});
+  //   }
+  //   return;
+  // }
```
Removed!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/827911/) on 8df46ac wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18203/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18203/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79d3c2f805a246394c54cd9eca37748c7d3dd5f`

lindsayad committed:
Fixups for CIVET

lindsayad commented:
This is looking really good as far as I'm concerned. My roadmap for merge:

- [ ] all recipes pass other than failures related to BISON
- [ ] I prepare patch for BISON
- [ ] Antonio approves the PR

moosebuild commented:
Job [Python 3.7 Tests](https://civet.inl.gov/job/831003/) on a2fa733 : invalidated by @lindsayad

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/831027/) on a2fa733 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b640e7</th>
      <th colspan=4>#18203 a2fa73</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.56%</td>
      <td>81.41%</td>
      <td>-0.15%</td>
      <td>66.35%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>69495</td>
      <td>69993</td>
      <td>+498</td>
      <td>562</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15716</td>
      <td>15982</td>
      <td>+266</td>
      <td>285</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18203/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18203/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>b640e7</th>
      <th colspan=4>#18203 a2fa73</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>78.15%</td>
      <td>77.71%</td>
      <td>-0.44%</td>
      <td>83.85%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2160</td>
      <td>2189</td>
      <td>+29</td>
      <td>109</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>604</td>
      <td>628</td>
      <td>+24</td>
      <td>21</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18203/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18203/coverage/contact)

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>b640e7</th>
      <th colspan=4>#18203 a2fa73</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.58%</td>
      <td>87.60%</td>
      <td>+0.03%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2622</td>
      <td>2621</td>
      <td>-1</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>372</td>
      <td>371</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18203/coverage_diff/heat_conduction.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18203/coverage/heat_conduction)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/rdg)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18203/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 66.35% is less than the suggested 90.0%
- `contact` new line coverage rate 83.85% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/831027/) on a2fa733 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 77.71% is less than the required 78.0%

lindsayad commented:
For the framework/moose-test, it looks like we're missing coverage for TRI6 and QUAD9 as well as debug printing.

jbadger95 commented:
@lindsayad The tests for TRI6 and QUAD9 were skipped because of minimum AD size, I put 200 on QUAD9 (might have been a bit overkill) but that one definitely needs more that the default 50. For the TRI6 test (continuity_tet10.i) I'm not actually sure it needs a bigger AD size, I think I modified that because I saw that it was segfaulting but that was related to the Libmesh bugs you fixed (I could be wrong though).

jbadger95 commented:
@lindsayad BTW, IIRC the sparsity size is a bit bigger than it needs to be, the sparsity makes room for the LM DoFs to be coupled to all the interior DoFs of secondary and primary elements when in reality the LM is only coupled to the interior DoFs that live on the mortar interface. Could probably reduce the AD size requirements quite a bit by not including that extra coupling.

lindsayad commented:
> Could probably reduce the AD size requirements quite a bit by not including that extra coupling.

I don't know how to do that in a smart and/or pretty way though :frowning_face:

lindsayad commented:
I suppose we could do a `if (_test[_i][_qp] == 0) continue;`  at the mortar constraint level

jbadger95 commented:
Yeah that'd be pretty simple, curious to know how much it reduces the AD sizes. Second order is where AD sizes really get crazy, this should reduce them by at least half.
